backend/src/apis/snapshot_data.rs

```diff
@@ -0,0 +1,312 @@
+use num_bigint::BigInt;
+use std::collections::HashMap;
+use std::{fs::File, io::BufReader};
+
+use halo2_proofs::{
+    halo2curves::bn256::{Bn256, Fr as Fp, G1Affine},
+    plonk::{keygen_pk, keygen_vk, ProvingKey, VerifyingKey},
+    poly::kzg::commitment::ParamsKZG,
+    SerdeFormat::RawBytes,
+};
+use snark_verifier_sdk::CircuitExt;
+
+use summa_solvency::{
+    circuits::{
+        merkle_sum_tree::MstInclusionCircuit,
+        solvency::SolvencyCircuit,
+        utils::{full_prover, generate_setup_params},
+    },
+    merkle_sum_tree::utils::big_int_to_fp,
+    merkle_sum_tree::MerkleSumTree,
+};
+
+use crate::apis::csv_parser::parse_csv_to_assets;
+
+struct MstParamsAndKeys<const LEVELS: usize, const L: usize, const N_ASSETS: usize> {
+    params: ParamsKZG<Bn256>,
+    pk: ProvingKey<G1Affine>,
+    vk: VerifyingKey<G1Affine>,
+}
+
+struct SnapshotData<
+    const LEVELS: usize,
+    const L: usize,
+    const N_ASSETS: usize,
+    const N_BYTES: usize,
+    const K: u32,
+> {
+    exchange_id: String,
+    mst: MerkleSumTree<N_ASSETS>,
+    assets: Vec<Asset>,
+    asset_signatures: AssetSignatures,
+    mst_params_and_keys: MstParamsAndKeys<LEVELS, L, N_ASSETS>,
+    proofs_of_inclusions: HashMap<u64, UserProof>,
+    proof_of_solvency: Option<SolvencyProof<N_ASSETS>>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Asset {
+    pub name: String,
+    pub pubkeys: Vec<String>,
+    pub balances: Vec<BigInt>,
+    pub sum_balances: Fp,
+}
+
+pub type AssetSignatures = HashMap<String, String>;
+
+#[derive(Debug, Clone)]
+struct UserProof {
+    // for each user
+    leaf_hash: Fp,
+    proof: Vec<u8>,
+}
+
+pub struct InclusionProof {
+    // public inputs
+    leaf_hash: Fp,
+    root_hash: Fp,
+    // need for verification
+    vk: VerifyingKey<G1Affine>,
+    pk: ProvingKey<G1Affine>,
+    proof: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+struct SolvencyProof<const N_ASSETS: usize> {
+    // public inputs
+    root_hash: Fp,
+    assets_sum: [Fp; N_ASSETS],
+    // need for verification
+    // vk is already exist in the onchain verifier
+    pk: ProvingKey<G1Affine>,
```

Comment:
Again, I don't think we need to store `pk` here, since it's never used

Reply:
Fixed at 71d4b19e5a87d3d236622d32bb9831840767be00